In [5]:
import tensorflow as tf
print("TensorFlow version:", tf.__version__)

import tensorflow_hub as hub
print("TensorFlow Hub version:", hub.__version__)

TensorFlow version: 2.17.0
TensorFlow Hub version: 0.16.1


In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as st
from scipy.stats import shapiro
import statsmodels.api as sm
import statsmodels.formula.api as smf
from datetime import datetime,date,timedelta
import datetime as dt
import time
from wordcloud import WordCloud # Visualisation
from nltk.tokenize import sent_tokenize, word_tokenize
from nltk.tokenize import wordpunct_tokenize
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
from nltk.corpus import words
from sklearn.metrics import coverage_error
from sklearn.metrics import label_ranking_average_precision_score
from sklearn.metrics import label_ranking_loss
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.multiclass import OneVsRestClassifier
import torch
import transformers as ppb
import warnings
warnings.filterwarnings('ignore')
from transformers import AutoTokenizer, DistilBertModel
import torch
import os
from sklearn.preprocessing import MultiLabelBinarizer

import mlflow
from mlflow.models import infer_signature

/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [7]:
df_net=pd.read_csv('df_net.csv', sep=',')
df_net

LastActivityDate                                              Title  \
0   2022-08-29 20:14:36                   Determine the type of an object?   
1   2022-07-02 05:29:08             Why can't Python parse this JSON data?   
2   2023-11-20 22:23:13   Best way to convert string to bytes in Python 3?   
3   2022-11-09 15:44:56                  Display number with leading zeros   
4   2023-07-28 17:04:02  Create a Pandas Dataframe by appending one row...   
5   2022-11-23 08:30:50  Should I use 'has_key()' or 'in' on Python dicts?   
6   2022-08-15 07:57:04      How to check if type of a variable is string?   
7   2023-01-27 10:02:02  Which exception should I raise on bad/illegal ...   
8   2022-05-31 17:20:40         Using @property versus getters and setters   
9   2023-01-29 17:03:17  How can I check if a string represents an int,...   
10  2023-08-23 15:00:24  How can I use pickle to save a dict (or any ot...   
11  2022-08-28 19:10:16     Getting a map() to return a list in Python 3.x   
12  2023-11-07 17:49:05                       How can I pivot a dataframe?   
13  2020-01-31 06:45:21         Determine if variable is defined in Python   
14  2022-07-05 21:20:38                           Reading JSON from a file   
15  2021-03-14 20:21:58                    Python Dictionary Comprehension   
16  2022-09-03 09:33:18  Is there a way to perform "if" in python's lam...   
17  2020-02-18 13:06:27         Convert a list of characters into a string   
18  2023-11-03 07:44:03  How to add to the PYTHONPATH in Windows, so it...   
19  2023-03-21 08:25:43                     How to find list intersection?   
20  2023-10-13 14:51:45               Redirect stdout to a file in Python?   
21  2021-06-25 14:13:28  Store output of subprocess.Popen call in a string   
22  2022-12-19 12:27:33        Why does range(start, end) not include end?   
23  2024-03-21 04:06:15         Convert a python dict to a string and back   
24  2022-11-02 08:35:13  How do you properly determine the current scri...   
25  2019-04-20 01:29:27   How can I check if a key exists in a dictionary?   
26  2022-08-06 19:04:52             Double Iteration in List Comprehension   
27  2022-03-30 14:18:48  What's the difference between `raw_input()` an...   
28  2022-05-24 08:21:27  NumPy array initialization (fill with identica...   
29  2021-06-28 09:57:35                  Python function global variables?   
30  2022-05-08 17:52:09                   Get loop count inside a for-loop   
31  2023-05-14 01:50:26  How do I convert all of the items in a list to...   
32  2022-12-29 23:24:40  'str' object has no attribute 'decode'. Python...   
33  2022-06-09 08:35:13  Access multiple elements of list knowing their...   
34  2022-04-05 10:52:38  Take the content of a list and append it to an...   
35  2023-02-02 18:17:15           List files ONLY in the current directory   
36  2023-04-24 15:57:39  Error "Import Error: No module named numpy" on...   
37  2023-12-28 20:34:37  Are nested try/except blocks in Python a good ...   
38  2015-11-06 10:04:38  How do I change the string representation of a...   
39  2024-02-13 18:27:19  Scraping: SSL: CERTIFICATE_VERIFY_FAILED error...   
40  2017-10-01 15:04:38          Python: access class property from string   
41  2022-07-30 13:07:53  How to count the frequency of the elements in ...   
42  2022-03-01 03:37:01  Find first element in a sequence that matches ...   
43  2021-07-13 19:58:45                  Confused by python file mode "w+"   
44  2022-05-01 16:39:08                          Python import csv to list   
45  2022-10-31 10:35:27                            Compute list difference   
46  2021-03-01 22:12:03    How do I pass a method as a parameter in Python   
47  2022-12-12 17:00:15  How to reliably open a file in the same direct...   
48  2021-02-09 11:47:31  ModuleNotFoundError: What does it mean __main_...   
49  2023-01-06 18:48:48     ImportError: No module named matplotlib.pyplot   

                            

In [8]:
# Load pre-trained universal sentence encoder model
embed = hub.load("https://tfhub.dev/google/universal-sentence-encoder/4")

In [9]:
# Sentences for which you want to create embeddings,
# passed as an array in embed()
list_sentences = df_net['sentence_dl'].tolist()
print(list_sentences)
embeddings = embed(df_net['sentence_dl'])
# Convertir les embeddings en numpy array pour un traitement ultérieur
embeddings_np = embeddings.numpy()

# Printing embeddings of each sentence
print(embeddings_np)

['determine the type of an object ?', 'why can t python parse this json data ?', 'best way to convert string to bytes in python 3 ?', 'display number with leading zeros', 'create a pandas dataframe by appending one row at a time', 'should i use has key ( ) or in on python dicts ?', 'how to check if type of a variable is string ?', 'which exception should i raise on bad illegal argument combinations in python ?', 'using property versus getters and setters', 'how can i check if a string represents an int , without using try except ?', 'how can i use pickle to save a dict ( or any other python object ) ?', 'getting a map ( ) to return a list in python 3.x', 'how can i pivot a dataframe ?', 'determine if variable is defined in python', 'reading json from a file', 'python dictionary comprehension', "is there a way to perform `` if '' in python s lambda ?", 'convert a list of characters into a string', 'how to add to the pythonpath in windows , so it finds my modules packages ?', 'how to fin

In [10]:
# To print each embeddings along with its corresponding
# sentence below code can be used.
for i in range(len(list_sentences)):
    print(list_sentences[i])
    print(embeddings[i])

determine the type of an object ?
tf.Tensor(
[ 0.00277023 -0.03878905 -0.02385747  0.01875251  0.03575515  0.06087856
  0.01388974 -0.01954049  0.07458097  0.07373869 -0.02220431  0.04348328
  0.02117297  0.0136591   0.02097735  0.08364316  0.01264817 -0.00398082
 -0.00199287 -0.02054167  0.09672691 -0.05919911  0.05339849  0.04130121
  0.01155577  0.00676485  0.01312432 -0.04181616  0.0643251  -0.0671135
 -0.03689624  0.0407103  -0.04543239  0.0020188  -0.05137011  0.01398868
 -0.0567608  -0.03268367 -0.02121354 -0.06381312  0.0344985   0.06881591
  0.00514044 -0.0563201   0.03849586 -0.01451128 -0.04036039 -0.00673036
  0.02748273 -0.06067921  0.06759673  0.10222576  0.02180594  0.01709348
 -0.01838033  0.02813969  0.01674365  0.07122819 -0.02331362 -0.07029837
 -0.02996477 -0.01700745 -0.03496189 -0.07376043  0.02777127 -0.05029924
 -0.00200418 -0.002424    0.05184152  0.01475967 -0.06066362  0.04782837
  0.05411956 -0.02786942 -0.08113559  0.0124495  -0.04144001 -0.04658065
 -0.010

In [11]:
tags = [tag[1:len(tag) - 1].split('><') for tag in df_net['Tags']]
tags

[['python', 'dictionary', 'types', 'typeof'],
 ['python', 'json', 'parsing'],
 ['python', 'string', 'character-encoding', 'python-3.x'],
 ['python', 'integer', 'string-formatting'],
 ['python', 'pandas', 'dataframe', 'append'],
 ['python', 'dictionary'],
 ['python', 'string', 'variables', 'types'],
 ['python', 'exception', 'arguments'],
 ['python', 'properties', 'getter-setter'],
 ['python', 'string', 'integer'],
 ['python', 'dictionary', 'pickle'],
 ['python', 'list', 'python-3.x', 'map-function'],
 ['python', 'pandas', 'group-by', 'pivot', 'pivot-table'],
 ['python', 'variables', 'defined'],
 ['python', 'json'],
 ['python', 'dictionary', 'list-comprehension'],
 ['python', 'lambda', 'conditional-operator'],
 ['python', 'string'],
 ['python', 'windows', 'environment-variables', 'pythonpath'],
 ['python', 'arrays', 'intersection'],
 ['python', 'stdout'],
 ['python', 'subprocess'],
 ['python', 'range'],
 ['python', 'json', 'dictionary', 'serialization'],
 ['python', 'path', 'working-dire

In [12]:
# Utilisation de MultiLabelBinarizer pour créer des étiquettes binaires
mlb = MultiLabelBinarizer()
labels = mlb.fit_transform(tags)
labels

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [13]:
# Séparation des données en ensembles d'entraînement et de test
train_title, test_title, train_labels, test_labels = train_test_split(embeddings_np, labels, test_size=0.2, random_state=42)
print("train_title shape : {}".format(train_title.shape))
print("test_title shape : {}".format(test_title.shape))
print("train_labels shape : {}".format(train_labels.shape))
print("test_labels shape : {}".format(test_labels.shape))

train_title shape : (40, 512)
test_title shape : (10, 512)
train_labels shape : (40, 66)
test_labels shape : (10, 66)


In [14]:
# Initialize Logistic Regression with OneVsRest
param_logit = {"estimator__C": [100, 10, 1.0, 0.1],
               "estimator__penalty": ["l1", "l2"],
               "estimator__dual": [False],
               "estimator__solver": ["liblinear"]}

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.model_selection import cross_val_score
from sklearn.multiclass import OneVsRestClassifier

lr = GridSearchCV(OneVsRestClassifier(LogisticRegression()),
                              param_grid=param_logit,
                              n_jobs=-1,
                              cv=5,
                              scoring="f1_weighted",
                              return_train_score = True,
                              refit=True)
lr.fit(train_title, train_labels)

# Predict
test_labels_predicted = lr.predict(test_title)

# Inverse transform
test_labels_pred_inversed = mlb\
    .inverse_transform(test_labels_predicted)
test_labels_inversed = mlb\
    .inverse_transform(test_labels)

print("Les 10 premiers tags prédits - vrais tags :")
print("Prédits:", test_labels_pred_inversed[0:10])
print("Vrais:", test_labels_inversed[0:10])

/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 3 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 4 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 6 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 9 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 11 is present in all training examples.
  warnings.

Les 10 premiers tags prédits - vrais tags :
Prédits: [('python',), ('python',), ('python',), ('list', 'python'), ('python',), ('python',), ('list', 'python', 'python-3.x'), ('python',), ('python',), ('list', 'python')]
Vrais: [('defined', 'python', 'variables'), ('beautifulsoup', 'python', 'scrapy', 'ssl-certificate', 'web-scraping'), ('for-loop', 'python'), ('list', 'python'), ('python', 'string'), ('module', 'python', 'python-3.6', 'python-import'), ('list-comprehension', 'python'), ('python',), ('python', 'python-3.x'), ('arrays', 'intersection', 'python')]


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 63 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 64 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 33 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 9 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: F-score is ill-defined and being

In [15]:
def metrics_score(model, df, y_true, y_pred):
    """Compilation function of metrics specific to multi-label
    classification problems in a Pandas DataFrame.
    This dataFrame will have 1 row per metric
    and 1 column per model tested.

    Exemples de Metriques
    ----------------------------------------
    coverage_error = 3 : En moyenne, il faut inclure trois prédictions pour capturer toutes les étiquettes pertinentes pour chaque instance.
    label_ranking_average_precision_score = 0.85 : La précision moyenne des classements est de 85%. Cela signifie que la majorité des étiquettes pertinentes sont bien classées en haut.
    label_ranking_loss = 0.1 : Il y a 10% d'inversions dans les classements où les étiquettes non pertinentes sont classées avant les étiquettes pertinentes.

    Parameters
    ----------------------------------------
    model : string
        Name of the tested model
    df : DataFrame
        DataFrame to extend.
        If None : Create DataFrame.
    y_true : array
        Array of true values to test
    y_pred : array
        Array of predicted values to test
    ----------------------------------------
    """
    if(df is not None):
        df_results = df
    else:
        df_results = pd.DataFrame(index=["coverage_error", "label_ranking_average_precision_score",
                                       "label_ranking_loss"],
                               columns=[model])

    scores = []
    scores.append(coverage_error(y_true,
                                         y_pred))
    scores.append(label_ranking_average_precision_score(y_pred,
                                   y_true))
    scores.append(label_ranking_loss(y_true,
                                       y_pred))

    df_results[model] = scores

    return df_results

In [16]:
df_metrics_compare_use = metrics_score("Logit",
                                   df=None,
                                   y_true = test_labels,
                                   y_pred = test_labels_predicted)
df_metrics_compare_use

Logit
coverage_error                         53.100000
label_ranking_average_precision_score   0.466212
label_ranking_loss                      0.490425

- coverage_error = 53.1 : Le modèle nécessite de parcourir un grand nombre de prédictions pour couvrir toutes les étiquettes pertinentes (53/66).
- label_ranking_average_precision_score = 0.466212 : La précision moyenne des classements des étiquettes est faible.
- label_ranking_loss = 0.490425 : Il y a 49% d'inversions dans les classements des étiquettes. C'est élevé.

Ce modèle donne des résultats meilleurs que BERT et Word2Vec mais cela reste faible.

In [17]:
# Initialiser le classificateur RandomForest
rf_model = RandomForestClassifier(random_state=42)
print("rf_model:", rf_model)

# Utiliser MultiOutputClassifier pour le modèle multi-label
rf_multioutput = OneVsRestClassifier(rf_model)
print("rf_multioutput:", rf_multioutput)

# Définir les hyperparamètres pour RandomizedSearchCV
param_rf = {
    "estimator__n_estimators": [50, 100, 200],
    "estimator__max_depth": [5,6,7],
    "estimator__min_samples_split": [2, 5, 10]
}

random_rf = RandomizedSearchCV(rf_multioutput, param_distributions=param_rf, cv=5, scoring='f1_weighted', n_jobs=-1)
random_rf.fit(train_title, train_labels)

# Meilleurs hyperparamètres
print("Meilleurs hyperparamètres trouvés par GridSearchCV:")
print(random_rf.best_params_)

# Prédictions avec le modèle optimisé
pred_labels = random_rf.predict(test_title)

# Inverse transform des prédictions optimisées
test_labels_pred_inversed_optimized = mlb.inverse_transform(pred_labels)
test_labels_inversed = mlb.inverse_transform(test_labels)

# Évaluer les performances avec le modèle optimisé
print("Les 10 premiers tags prédits avec le modèle optimisé - vrais tags :")
print("Prédits:", test_labels_pred_inversed_optimized[0:10])
print("Vrais:", test_labels_inversed[0:10])

df_metrics_compare_use_rf = metrics_score("Random Forest",
                                   df=None,
                                   y_true = test_labels,
                                   y_pred = pred_labels)

df_metrics_compare_use_rf

rf_model: RandomForestClassifier(random_state=42)
rf_multioutput: OneVsRestClassifier(estimator=RandomForestClassifier(random_state=42))


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 0 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 0 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 1 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 2 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 3 is present in all training examples.
  warnings.w

Meilleurs hyperparamètres trouvés par GridSearchCV:
{'estimator__n_estimators': 100, 'estimator__min_samples_split': 10, 'estimator__max_depth': 5}
Les 10 premiers tags prédits avec le modèle optimisé - vrais tags :
Prédits: [('python',), ('python',), ('python',), ('python',), ('python',), ('python',), ('python',), ('python',), ('python',), ('python',)]
Vrais: [('defined', 'python', 'variables'), ('beautifulsoup', 'python', 'scrapy', 'ssl-certificate', 'web-scraping'), ('for-loop', 'python'), ('list', 'python'), ('python', 'string'), ('module', 'python', 'python-3.6', 'python-import'), ('list-comprehension', 'python'), ('python',), ('python', 'python-3.x'), ('arrays', 'intersection', 'python')]


Random Forest
coverage_error                             59.500000
label_ranking_average_precision_score       0.461667
label_ranking_loss                          0.538333

In [18]:
# Initialiser le classificateur GradientBoostingClassifier
gb_model = GradientBoostingClassifier(random_state=42)
gb_multioutput = OneVsRestClassifier(gb_model, n_jobs=-1)

# Entraînement du modèle
gb_multioutput.fit(train_title, train_labels)
print("Modèle entraîné.")

# Prédictions
pred_labels = gb_multioutput.predict(test_title)

# Inverse transform des prédictions
test_labels_pred_inversed = mlb.inverse_transform(pred_labels)
test_labels_inversed = mlb.inverse_transform(test_labels)

# Évaluer les performances avec le modèle optimisé
print("Les 10 premiers tags prédits avec le modèle optimisé - vrais tags :")
print("Prédits:", test_labels_pred_inversed_optimized[0:10])
print("Vrais:", test_labels_inversed[0:10])


df_metrics_compare_use_gb = metrics_score("Gradient Boosting",
                                   df=None,
                                   y_true = test_labels,
                                   y_pred = pred_labels)
df_metrics_compare_use_gb

/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 3 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 9 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 16 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 25 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 33 is present in all training examples.
  warning

Modèle entraîné.
Les 10 premiers tags prédits avec le modèle optimisé - vrais tags :
Prédits: [('python',), ('python',), ('python',), ('python',), ('python',), ('python',), ('python',), ('python',), ('python',), ('python',)]
Vrais: [('defined', 'python', 'variables'), ('beautifulsoup', 'python', 'scrapy', 'ssl-certificate', 'web-scraping'), ('for-loop', 'python'), ('list', 'python'), ('python', 'string'), ('module', 'python', 'python-3.6', 'python-import'), ('list-comprehension', 'python'), ('python',), ('python', 'python-3.x'), ('arrays', 'intersection', 'python')]


Gradient Boosting
coverage_error                                 53.100000
label_ranking_average_precision_score           0.457879
label_ranking_loss                              0.490677

In [19]:
pd.concat([df_metrics_compare_use, df_metrics_compare_use_rf, df_metrics_compare_use_gb], axis=1)

Logit  Random Forest  \
coverage_error                         53.100000      59.500000   
label_ranking_average_precision_score   0.466212       0.461667   
label_ranking_loss                      0.490425       0.538333   

                                       Gradient Boosting  
coverage_error                                 53.100000  
label_ranking_average_precision_score           0.457879  
label_ranking_loss                              0.490677

Je décide de travailler avec la Régression Logistique qui me donne les meilleurs résultats.


In [21]:
# Configurer les avertissements et MLflow
os.environ["TOKENIZERS_PARALLELISM"] = "false"
warnings.filterwarnings("ignore", message="Label not .* is present in all training examples", category=UserWarning, module="sklearn")

Results_USE = pd.DataFrame(index=['coverage_error', 'coverage_error_optimized', 'label_ranking_average_precision_score',
                                  'label_ranking_average_precision_score_optimized', 'label_ranking_loss', 'label_ranking_loss_optimized', 
                                  'Training Time', 'Training Random Time', 'Testing Time', 'Testing Random Time'],
                           columns=['LogisticRegression', 'RandomForestClassifier', 'GradientBoostingClassifier'])

# Hyperparamètres
lr_params = {"estimator__C": [100, 10, 1.0, 0.1],
             "estimator__penalty": ["l1", "l2"],
             "estimator__dual": [False],
             "estimator__solver": ["liblinear"]}

random_forest_params = {
    "estimator__n_estimators": [50, 100, 200],
    "estimator__max_depth": [5, 6, 7],
    "estimator__min_samples_split": [2, 5, 10]}

gb_params = {
    'estimator__n_estimators': [100, 200, 300],
    'estimator__learning_rate': [0.01, 0.05, 0.1],
    'estimator__max_depth': [3, 4, 5]}

myrandcvs = [lr_params, random_forest_params, gb_params]

# Modèles
classifiers = {'LogisticRegression': LogisticRegression(), 'RandomForestClassifier': RandomForestClassifier(), 'GradientBoostingClassifier': GradientBoostingClassifier()}

mlflow.set_tracking_uri("http://127.0.0.1:8080")

# Définir une nouvelle expérience MLflow
mlflow.set_experiment("Multilabel_Classification_Experiment_USE")

for model_name, model in classifiers.items():
    model_ovr = OneVsRestClassifier(model)
    randcv = RandomizedSearchCV(estimator=model_ovr, param_distributions=myrandcvs[0 if model_name == 'LogisticRegression'
                                                                                   else 1 if model_name == 'RandomForestClassifier'
                                                                                   else 2], n_iter=5, cv=3, n_jobs=-1, verbose=2, scoring='accuracy', random_state=42)

    # Démarrer une nouvelle exécution MLflow
    with mlflow.start_run(run_name=model_name):
        # Entraînement
        start_time_fit = time.time()
        model_ovr.fit(train_title, train_labels)
        time_fit = (time.time() - start_time_fit)

        # Random Training
        start_rdcv_time_fit = time.time()
        randcv.fit(train_title, train_labels)
        time_rdcv_fit = (time.time() - start_rdcv_time_fit)

        # Test
        start_time_predict = time.time()
        y_pred = model_ovr.predict(test_title)
        time_predict = (time.time() - start_time_predict)

        # Random Test
        start_rdcv_time_predict = time.time()
        y_rdcv_pred = randcv.predict(test_title)
        time_rdcv_predict = (time.time() - start_rdcv_time_predict)

        # Métriques
        coverage_errors = coverage_error(test_labels, y_pred)
        label_ranking_average_precision_scores = label_ranking_average_precision_score(test_labels, y_pred)
        label_ranking_losses = label_ranking_loss(test_labels, y_pred)

        # Métriques Random
        coverage_error_optimized = coverage_error(test_labels, y_rdcv_pred)
        label_ranking_average_precision_score_optimized = label_ranking_average_precision_score(test_labels, y_rdcv_pred)
        label_ranking_loss_optimized = label_ranking_loss(test_labels, y_rdcv_pred)

        # Enregistrer les hyperparamètres, les métriques et le modèle dans MLflow
        mlflow.log_params(randcv.best_params_)
        mlflow.log_metric("coverage_error", coverage_errors)
        mlflow.log_metric("coverage_error_optimized", coverage_error_optimized)
        mlflow.log_metric("label_ranking_average_precision_score", label_ranking_average_precision_scores)
        mlflow.log_metric("label_ranking_average_precision_score_optimized", label_ranking_average_precision_score_optimized)
        mlflow.log_metric("label_ranking_loss", label_ranking_losses)
        mlflow.log_metric("label_ranking_loss_optimized", label_ranking_loss_optimized)
        mlflow.log_metric("Training Time", time_fit)
        mlflow.log_metric("Training Random Time", time_rdcv_fit)
        mlflow.log_metric("Testing Time", time_predict)
        mlflow.log_metric("Testing Random Time", time_rdcv_predict)
        mlflow.sklearn.log_model(randcv.best_estimator_, "model")

        # Mise à jour du DataFrame
        Results_USE.loc['coverage_error', model_name] = coverage_errors
        Results_USE.loc['coverage_error_optimized', model_name] = coverage_error_optimized
        Results_USE.loc['label_ranking_average_precision_score', model_name] = label_ranking_average_precision_scores
        Results_USE.loc['label_ranking_average_precision_score_optimized', model_name] = label_ranking_average_precision_score_optimized
        Results_USE.loc['label_ranking_loss', model_name] = label_ranking_losses
        Results_USE.loc['label_ranking_loss_optimized', model_name] = label_ranking_loss_optimized
        Results_USE.loc['Training Time', model_name] = time_fit
        Results_USE.loc['Training Random Time', model_name] = time_rdcv_fit
        Results_USE.loc['Testing Time', model_name] = time_predict
        Results_USE.loc['Testing Random Time', model_name] = time_rdcv_predict

Fitting 3 folds for each of 5 candidates, totalling 15 fits


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 1 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 2 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 3 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 0 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 0 is present in all training examples.
  warnings.w

[CV] END estimator__C=1.0, estimator__dual=False, estimator__penalty=l2, estimator__solver=liblinear; total time=   0.1s
[CV] END estimator__C=100, estimator__dual=False, estimator__penalty=l2, estimator__solver=liblinear; total time=   0.1s
[CV] END estimator__C=100, estimator__dual=False, estimator__penalty=l2, estimator__solver=liblinear; total time=   0.1s
[CV] END estimator__C=100, estimator__dual=False, estimator__penalty=l1, estimator__solver=liblinear; total time=   0.1s
[CV] END estimator__C=1.0, estimator__dual=False, estimator__penalty=l2, estimator__solver=liblinear; total time=   0.1s
[CV] END estimator__C=100, estimator__dual=False, estimator__penalty=l2, estimator__solver=liblinear; total time=   0.1s
[CV] END estimator__C=1.0, estimator__dual=False, estimator__penalty=l2, estimator__solver=liblinear; total time=   0.1s
[CV] END estimator__C=100, estimator__dual=False, estimator__penalty=l1, estimator__solver=liblinear; total time=   0.1s
[CV] END estimator__C=0.1, estim

2024/08/22 16:19:47 WARNING mlflow.models.model: Input example should be provided to infer model signature if the model signature is not provided when logging the model.
2024/08/22 16:19:47 INFO mlflow.tracking._tracking_service.client: 🏃 View run LogisticRegression at: http://127.0.0.1:8080/#/experiments/415837045020450816/runs/45fbfdd74aea4c939d6e5fa6575f6026.
2024/08/22 16:19:47 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:8080/#/experiments/415837045020450816.


Fitting 3 folds for each of 5 candidates, totalling 15 fits


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 0 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 0 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 0 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 0 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 1 is present in all training examples.
  warnings.w

[CV] END estimator__max_depth=6, estimator__min_samples_split=2, estimator__n_estimators=50; total time=   1.3s
[CV] END estimator__max_depth=7, estimator__min_samples_split=5, estimator__n_estimators=50; total time=   1.2s
[CV] END estimator__max_depth=6, estimator__min_samples_split=2, estimator__n_estimators=50; total time=   1.4s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 9 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 40 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 41 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 3 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 9 is present in all training examples.
  warnings

[CV] END estimator__max_depth=6, estimator__min_samples_split=2, estimator__n_estimators=50; total time=   1.5s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 11 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 21 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 22 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 23 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 16 is present in all training examples.
  warni

[CV] END estimator__max_depth=6, estimator__min_samples_split=5, estimator__n_estimators=100; total time=   2.9s
[CV] END estimator__max_depth=6, estimator__min_samples_split=5, estimator__n_estimators=100; total time=   3.1s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 59 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 54 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 35 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 36 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 62 is present in all training examples.
  warni

[CV] END estimator__max_depth=7, estimator__min_samples_split=5, estimator__n_estimators=50; total time=   1.9s
[CV] END estimator__max_depth=5, estimator__min_samples_split=2, estimator__n_estimators=50; total time=   1.7s
[CV] END estimator__max_depth=7, estimator__min_samples_split=5, estimator__n_estimators=50; total time=   1.9s
[CV] END estimator__max_depth=6, estimator__min_samples_split=5, estimator__n_estimators=100; total time=   3.2s
[CV] END estimator__max_depth=5, estimator__min_samples_split=2, estimator__n_estimators=50; total time=   1.8s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 37 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 16 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 40 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 41 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 25 is present in all training examples.
  warni

[CV] END estimator__max_depth=5, estimator__min_samples_split=2, estimator__n_estimators=50; total time=   1.1s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 42 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 43 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 57 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 51 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 52 is present in all training examples.
  warni

[CV] END estimator__max_depth=5, estimator__min_samples_split=10, estimator__n_estimators=200; total time=   4.8s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 63 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 64 is present in all training examples.
  warnings.warn(


[CV] END estimator__max_depth=5, estimator__min_samples_split=10, estimator__n_estimators=200; total time=   5.2s
[CV] END estimator__max_depth=5, estimator__min_samples_split=10, estimator__n_estimators=200; total time=   5.3s


2024/08/22 16:20:02 WARNING mlflow.models.model: Input example should be provided to infer model signature if the model signature is not provided when logging the model.
2024/08/22 16:20:02 INFO mlflow.tracking._tracking_service.client: 🏃 View run RandomForestClassifier at: http://127.0.0.1:8080/#/experiments/415837045020450816/runs/83db7c2993464690bc82d25eb92d40f6.
2024/08/22 16:20:02 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:8080/#/experiments/415837045020450816.


Fitting 3 folds for each of 5 candidates, totalling 15 fits


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 0 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 0 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 0 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 0 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 1 is present in all training examples.
  warnings.w

[CV] END estimator__learning_rate=0.05, estimator__max_depth=3, estimator__n_estimators=100; total time=   4.0s
[CV] END estimator__learning_rate=0.1, estimator__max_depth=4, estimator__n_estimators=100; total time=   4.0s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 25 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 26 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 60 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 3 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 4 is present in all training examples.
  warning

[CV] END estimator__learning_rate=0.05, estimator__max_depth=3, estimator__n_estimators=100; total time=   4.6s
[CV] END estimator__learning_rate=0.05, estimator__max_depth=3, estimator__n_estimators=100; total time=   4.8s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 3 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 44 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label 45 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 46 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 15 is present in all training examples.
  warnings.w

[CV] END estimator__learning_rate=0.05, estimator__max_depth=4, estimator__n_estimators=200; total time=   7.9s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 54 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 58 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 52 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 53 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 54 is present in all training examples.
  warni

[CV] END estimator__learning_rate=0.1, estimator__max_depth=4, estimator__n_estimators=100; total time=   4.6s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 63 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 64 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 62 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 63 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 65 is present in all training examples.
  warni

[CV] END estimator__learning_rate=0.1, estimator__max_depth=4, estimator__n_estimators=100; total time=   4.8s
[CV] END estimator__learning_rate=0.01, estimator__max_depth=3, estimator__n_estimators=100; total time=   4.1s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 16 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 60 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 63 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 64 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label 45 is present in all training examples.
  warnings.

[CV] END estimator__learning_rate=0.05, estimator__max_depth=4, estimator__n_estimators=200; total time=   9.0s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 63 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 52 is present in all training examples.
  warnings.warn(


[CV] END estimator__learning_rate=0.05, estimator__max_depth=4, estimator__n_estimators=200; total time=   9.3s
[CV] END estimator__learning_rate=0.01, estimator__max_depth=3, estimator__n_estimators=100; total time=   4.5s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 48 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 54 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 25 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 39 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 51 is present in all training examples.
  warni

[CV] END estimator__learning_rate=0.01, estimator__max_depth=5, estimator__n_estimators=300; total time=  11.5s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 63 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 64 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 63 is present in all training examples.
  warnings.warn(


[CV] END estimator__learning_rate=0.01, estimator__max_depth=3, estimator__n_estimators=100; total time=   4.0s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 59 is present in all training examples.
  warnings.warn(


[CV] END estimator__learning_rate=0.01, estimator__max_depth=5, estimator__n_estimators=300; total time=  12.4s


/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 63 is present in all training examples.
  warnings.warn(
/Users/randaalsabbagh/Desktop/MACHINE_LEARNING/P5/Projet-5-ML/envp5/lib/python3.9/site-packages/sklearn/multiclass.py:87: UserWarning: Label not 64 is present in all training examples.
  warnings.warn(


[CV] END estimator__learning_rate=0.01, estimator__max_depth=5, estimator__n_estimators=300; total time=  13.2s


2024/08/22 16:20:32 WARNING mlflow.models.model: Input example should be provided to infer model signature if the model signature is not provided when logging the model.
2024/08/22 16:20:32 INFO mlflow.tracking._tracking_service.client: 🏃 View run GradientBoostingClassifier at: http://127.0.0.1:8080/#/experiments/415837045020450816/runs/7671c3ffec194848b0d05f2167e9a4fd.
2024/08/22 16:20:32 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:8080/#/experiments/415837045020450816.


In [22]:
Results_USE.index.name = 'Metrics USE'
# Renommer la première colonne
Results_USE.rename(columns={'index': 'Metrics USE'}, inplace=False)

Results_USE

LogisticRegression  \
Metrics USE                                                          
coverage_error                                                59.5   
coverage_error_optimized                                      59.5   
label_ranking_average_precision_score                     0.485909   
label_ranking_average_precision_score_optimized           0.469242   
label_ranking_loss                                        0.538333   
label_ranking_loss_optimized                              0.538862   
Training Time                                             1.537051   
Training Random Time                                      1.706705   
Testing Time                                               0.00226   
Testing Random Time                                       0.002003   

                                                RandomForestClassifier  \
Metrics USE                                                              
coverage_error                                                    59.5   
coverage_error_optimized                                          59.5   
label_ranking_average_precision_score                         0.485909   
label_ranking_average_precision_score_optimized               0.485909   
label_ranking_loss                                            0.538333   
label_ranking_loss_optimized                                  0.538333   
Training Time                                                 2.710147   
Training Random Time                                         10.621898   
Testing Time                                                  0.088567   
Testing Random Time                                           0.168014   

                                                GradientBoostingClassifier  
Metrics USE                                                                 
coverage_error                                                        53.1  
coverage_error_optimized                                              53.1  
label_ranking_average_precision_score                             0.476061  
label_ranking_average_precision_score_optimized                   0.476061  
label_ranking_loss                                                0.490677  
label_ranking_loss_optimized                                      0.490677  
Training Time                                                     7.525193  
Training Random Time                                             20.698212  
Testing Time                                                      0.007041  
Testing Random Time                                                0.00678

In [23]:
import joblib
# Sauvegarder le modèle dans un fichier .pkl
joblib.dump(model, 'use_model.pkl')

['use_model.pkl']